### PR TITLE
Iteration 4

### PR DESCRIPTION
**Pull request by gratefulcheddar**
[Quit/Restart functionality is complete](https://trello.com/c/OmQo3kaS/8-quit-and-restart-game-options-2) 
[README File](https://trello.com/c/w4OtFvHN/19-create-readme-file-1) 
[Returns a list of input that was invalid for each guess](https://trello.com/c/lltU3MWf/12-add-details-to-guess-feedback-2) 

(Links go to corresponding Trello cards/stories)

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1 +1,22 @@
...
+
+## How To Install
+
+Ruby Version 2.5.1
+
+Install bundler gem, clone repo, run the bundle install command within the project directory to install dependencies.
+
+```
+gem install bundler
+git clone https://github.com/gratefulcheddar/mastermind-8thlight
```
Comment (ssunday) [inline README.md]: Nice instructions!

File: README.md [hunk 2]
```diff
@@ -1 +1,22 @@
-MasterMind
+# MasterMind
+
+## How To Install
+
+Ruby Version 2.5.1
```
Comment (ssunday) [inline README.md]: Maybe list this under a requirements section (which you would also put bundler)?

File: lib/mastermind_io.rb [hunk 3]
```diff
@@ -44,7 +44,9 @@ def get_guess(correct_length)
 
       if guess.count == correct_length
         return guess if Mastermind.validate_colors(guess)
-        output @messages.invalid_color_error_message
+        invalid_colors = guess.reject { |color| Mastermind::COLOR_OPTIONS.include? color}
+        invalid_colors.map! { |color| color.to_s }
```
Comment (ssunday) [inline lib/mastermind_io.rb]: Did you try doing `invalid_colors.map!(&:to_s)`? With map, you can pass `&:method_name` to collect the results of calling a method on a list of objects/things.

File: lib/main.rb [hunk 4]
```diff
@@ -1,4 +1,25 @@
...
+    replay = gets.chomp.downcase.to_sym
+    while replay != :no && replay != :yes
+      print "\nEnter yes or no: "
+      replay = gets.chomp.downcase.to_sym
+    end
+    break if replay == :yes
+  end
+  print "Would you like to play again? (Enter yes or no): "
+  replay = gets.chomp.downcase.to_sym
+  while replay != :no && replay != :yes
```
Comment (ssunday) [inline lib/main.rb]: This code is repeated...could you make a little function to replace it ?

File: lib/mastermind_dialog.rb [hunk 5]
```diff
@@ -11,12 +11,16 @@ def color_input_message
     "Guess 4 colors (red, blue, green, orange, yellow, or purple) separated by spaces and press Enter"
   end
 
-  def invalid_color_error_message
-    "Invalid color option entered. Please guess again.\n"
+  def invalid_color_error_message(invalid_colors)
+    if invalid_colors.count == 1
+      "#{invalid_colors} is not a valid color. Please guess again. \n"
```
Comment (ssunday) [inline lib/mastermind_dialog.rb]: You could in theory have the ternary inside the string formatting, might be cleaner

like `#{invalid_colors.count == 1 ? 'is' : 'are'}`

File: lib/mastermind_io.rb (hunk 3 above)
Comment (gratefulcheddar) [reply lib/mastermind_io.rb]: Ah yes! I use that somewhere else and forgot about it here. Will change, thanks!

File: lib/main.rb [hunk 6]
```diff
@@ -5,19 +5,11 @@
...
+    replay = MastermindIO.new.get_yes_no_answer
     break if replay == :yes
   end
   print "Would you like to play again? (Enter yes or no): "
-  replay = gets.chomp.downcase.to_sym
-  while replay != :no && replay != :yes
-    print "\nEnter yes or no: "
-    replay = gets.chomp.downcase.to_sym
-  end
+  replay = MastermindIO.new.get_yes_no_answer
```
Comment (ssunday) [inline lib/main.rb]: If you're just instantiating it to call a method, maybe make it a static method so you don't have to do .new.

**Pull request merged**